test/python/transpiler/test_consolidate_blocks.py

```diff
@@ -146,6 +149,61 @@ def test_block_spanning_two_regs(self):
         fidelity = process_fidelity(new_dag.op_nodes()[0].op.to_matrix(), unitary.to_matrix())
         self.assertAlmostEqual(fidelity, 1.0, places=7)
 
+    def test_node_added_before_block(self):
+        """Test that a node before a block remains before the block"""
+        qc = QuantumCircuit(2, 1)
+        qc.iden(0)
+        qc.swap(1, 0)
+        qc.cx(1, 0)
+        qc.iden(1)
+
+        # can't just add all the nodes to one block as in other tests
+        # as we are trying to test the block gets added in the correct place
+        # so use a pass to collect the blocks instead
+        pass_manager = PassManager()
+        pass_manager.append(Collect2qBlocks())
```

Comment:
I think `Collect2qBlocks` assumes the circuit has already been unrolled to `['cx', 'u1', 'u2', 'u3']`. (Maybe that's why we didn't initially see the issue in #2775 .)

Reply:
I think that `id` is [also allowed](https://github.com/Qiskit/qiskit-terra/blob/a444801be9f8d1f2657eac2778541edb129e47e9/qiskit/transpiler/passes/collect_2q_blocks.py#L51), but yes that could be the issue there. I will have a look into it.

Reply:
I have updated the code so that #2775 is now also fixed by this PR. `Collect2qBlocks` didn't expect to have measures in the middle of the circuit as you said, it was only looking for a subset of gates. I have updated it so that these should now work and added in the test from #2775.